### PR TITLE
add mp4 muxer.

### PR DESCRIPTION
**Pull request by xuguangxin**
@kodabb and @tdaede 
This will fix  https://github.com/xiph/rav1e/issues/998.
I am not familiar with rust. Any suggestion is welcome.

Please ignore de2611a and 32b57af, I already sent them out at https://github.com/xiph/rav1e/pull/1259

thanks

### CODE REVIEW

**Comment by xuguangxin** [comment]
to test this feature. you need run "cargo run --features="ffmpeg-sys" in.y4m --output out.mp4"

File: src/bin/muxer.rs [hunk 1]
```diff
@@ -153,6 +163,129 @@ impl IvfMuxer {
   }
 }
 
+#[cfg(feature = "ffmpeg-sys")]
```
[the remark this answers is not in the document]
Comment (xuguangxin) [reply src/bin/muxer.rs]: thanks for the suggestion. I will prepare the patch.

**Comment by xuguangxin** [comment]
@lu-zero fixed, please help review again. 
thanks

File: src/bin/mux/mod.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#[cfg(feature = "ffmpeg-sys")]
+mod avformatmuxer;
+#[cfg(feature = "ffmpeg-sys")]
+use avformatmuxer::AvformatMuxer;
+
+use std::ffi::OsStr;
+use std::path::Path;
+
+fn need_container(path: &str) -> bool {
+  let ext =
```
Comment (lu-zero) [inline src/bin/mux/mod.rs]: This function is still wrong. You do not want to produce an `ivf`  if the extension is not supported.

File: src/bin/mux/avformatmuxer.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+// the Alliance for Open Media Patent License 1.0. If the BSD 2 Clause License
+// was not distributed with this source code in the LICENSE file, you can
+// obtain it at www.aomedia.org/license/software. If the Alliance for Open
+// Media Patent License 1.0 was not distributed with this source code in the
+// PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+
+#[cfg(feature = "ffmpeg-sys")]
+use ffmpeg_sys::*;
+
+#[allow(unused_imports)]
```
Comment (lu-zero) [inline src/bin/mux/avformatmuxer.rs]: Why this?

File: src/bin/mux/mod.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      //need time to figure out.
+      "mp4" => true,
+      _ => false
+    },
+    _ => false
+  }
+}
+
+#[allow(unreachable_code)]
+pub fn create_muxer(path: &str) -> Box<dyn Muxer> {
```
Comment (lu-zero) [inline src/bin/mux/mod.rs]: This should panic if the extension isn't supported.

File: src/bin/rav1e.rs [hunk 5]
```diff
@@ -7,16 +7,13 @@
...
 
-use y4m;
-
 #[macro_use]
 extern crate scan_fmt;
 
 mod common;
 mod decoder;
-mod muxer;
+mod mux;
```
Comment (lu-zero) [inline src/bin/rav1e.rs]: I wouldn't change the name

File: src/bin/mux/mod.rs (hunk 2 above)
Comment (xuguangxin) [reply src/bin/mux/mod.rs]: ok, I will change to this

File: src/bin/rav1e.rs (hunk 5 above)
Comment (xuguangxin) [reply src/bin/rav1e.rs]: We put the Muxer trait to muxer.rc, clippy will report an error if the directory also named muxer. 
any suggestion for fix this?
thanks

**Comment by xuguangxin** [comment]
@lu-zero , everything changed except the mux namespace. 
thanks

**Comment by xuguangxin** [comment]
rebased for https://github.com/xiph/rav1e/commit/bb8d569030b691a3f003e1133647484b98b467be, 
@kodabb , any comment for this pr?

thanks

**Comment by kodabb** [comment]
it's on my list of code to review, i'll get to it this week

**Comment by xuguangxin** [comment]
sure, take your time.
thanks

File: src/bin/rav1e.rs (hunk 5 above)
Comment (lu-zero) [reply src/bin/rav1e.rs]: The layout I had in mind is:
```
|- ./
   |- decoder/
      |- mod.rs
      |- y4m.rs
   |- muxer/
      |- ivf.rs
      |- mod.rs
      |- mp4.rs
      |- mkv.rs
   |- rav1e.rs
   |- common.rs
```

I just checked and this seems working :)

Comment (xuguangxin) [reply src/bin/rav1e.rs]: two questions for this. 
1. avformat can support many extensions like .mp4,  .mkv, .webm. Do we really need different files for mp4 and mkv? What is the benefit?
2.  We have two major things in the muxer module. The muxer trait and the create_muxer factory function.  Where we put these two things?
Put them all in mod.rs may introduce cycle dependency. ivf and mp4 depend on mods because of Muxer trait.  mods depend on ivf, mp4 because of create_muxer.

thanks

Comment (lu-zero) [reply src/bin/rav1e.rs]: > two questions for this.
> 
>     1. avformat can support many extensions like .mp4,  .mkv, .webm. Do we really need different files for mp4 and mkv? What is the benefit?

We do.

Avformat is not the only provider for muxing, e.g. l-smash is traditionally preferred to mux mp4 when available.

>     2. We have two major things in the muxer module. The muxer trait and the create_muxer factory function.  Where we put these two things?
>        Put them all in mod.rs may introduce cycle dependency. ivf and mp4 depend on mods because of Muxer trait.  mods depend on ivf, mp4 because of create_muxer.

There aren't problems with cycle dependencies in rust.

Comment (xuguangxin) [reply src/bin/rav1e.rs]: good to know those, I will update according to your comment.
>Avformat is not the only provider for muxing, e.g. l-smash is traditionally preferred to mux mp4 when available.

> There aren't problems with cycle dependencies in rust.

thanks for suggestion

Comment (lu-zero) [reply src/bin/rav1e.rs]: Thank you for your work :) I know how frustrating can be refining a patch for something that works well enough for you already.

Comment (xuguangxin) [reply src/bin/rav1e.rs]: Do not worry. Your request is very reasonable. The way is more clear than my original patch. I am happy to do this.

Comment (xuguangxin) [reply src/bin/rav1e.rs]: updated, please help review.
thanks

**Comment by kodabb** [comment]
> thread 'main' panicked at 'need ffmpeg-sys for .mp4, please build with --features="ffmpeg-sys", or you can use .ivf extesion', src/bin/muxer/mp4.rs:26:7

should this module be enabled by default? i think that the error message could be scary but at least it explains how to do it which i like

>      handler_name    : VideoHandler

is it possible to set the name of the encoder to `rav1e` + version instead of the default videoHandler? or maybe there is a more specific place where this could be added

anyway all my major points were already covered by @lu-zero, so aside my two questions i think it looks good!

thanks for providing this functionality

**Comment by xuguangxin** [comment]
> thread 'main' panicked at 'need ffmpeg-sys for .mp4, please build with --features="ffmpeg-sys", or you can use .ivf extesion', src/bin/muxer/mp4.rs:26:7
> 
> should this module be enabled by default? i think that the error message could be scary but at least it explains how to do it which i like
> 

ffmpeg-sys need ffmpeg dev env, and it may only work with the latest version of ffmpeg.
Enable it by default may introduce trouble if somebody does not care about the mp4.
I can enable it in travis if you think it's ok.


>  handler_name    : VideoHandler
> is it possible to set the name of the encoder to rav1e + version instead of the default videoHandler? or maybe there is a more specific place where this could be added
> 

This is is a new requirement. I can create and follow up it with a new issue.

> anyway all my major points were already covered by @lu-zero, so aside my two questions i think it looks good!
> 
> thanks for providing this functionality

welcome, I am willing to do any thing benefit the av1 eco system.

**Comment by xuguangxin** [comment]
@lu-zero , @kodabb , any more comments on this? Can we merge it now?

thanks

**Comment by xuguangxin** [comment]
Welcome, thanks for careful review.  Constructive feedback from you and kodabb helps me a lot.

**Comment by kodabb** [comment]
@lu-zero all good on my side

File: src/bin/muxer/avformat.rs [hunk 6]
```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2017-2018, The rav1e contributors. All rights reserved
```
Comment (dwbuiten) [inline src/bin/muxer/avformat.rs]: Wrong copyright year?

File: src/bin/muxer/avformat.rs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+      match avformat_alloc_output_context2(
+        &mut context,
+        ptr::null_mut(),
+        ptr::null(),
+        p.as_ptr() as *const i8
+      ) {
+        0 => {
+          let mut io = ptr::null_mut();
+          let ret =
+            avio_open(&mut io, p.as_ptr() as *const i8, AVIO_FLAG_WRITE);
```
Comment (dwbuiten) [inline src/bin/muxer/avformat.rs]: Why are we using AVIO directly?

File: src/bin/muxer/avformat.rs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+      let framerate_num = framerate_num as i32;
+      let duration = (*stream).time_base.den * framerate_den
+        / ((*stream).time_base.num * framerate_num);
+      self.duration = duration as i64;
+    }
+  }
+
+  fn write_frame(&mut self, pts: u64, data: &[u8]) {
+    unsafe {
+      let mut pkt: AVPacket = mem::zeroed();
```
Comment (dwbuiten) [inline src/bin/muxer/avformat.rs]: `av_packet_alloc()`?

File: src/bin/muxer/avformat.rs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+      (*param).height = height as i32;
+      let ret = avformat_write_header(self.context, ptr::null_mut());
+      if ret < 0 {
+        panic!("write header failed error = {}", ret);
+      }
+      let framerate_den = framerate_den as i32;
+      let framerate_num = framerate_num as i32;
+      let duration = (*stream).time_base.den * framerate_den
+        / ((*stream).time_base.num * framerate_num);
+      self.duration = duration as i64;
```
Comment (dwbuiten) [inline src/bin/muxer/avformat.rs]: Since this looks to be used for PTS rescaling, storing this as an integer and just multiplying it later seems not-so-accurate (?). Is there a reason not to use proper rational scaling?

File: src/bin/muxer/avformat.rs [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    unsafe {
+      let mut pkt: AVPacket = mem::zeroed();
+      av_init_packet(&mut pkt);
+      pkt.data = data.as_ptr() as *mut _;
+      pkt.size = data.len() as i32;
+
+      let pts = pts as i64;
+      pkt.pts = pts * self.duration;
+      //no b frame, pts equals dts
+      pkt.dts = pts * self.duration;
```
Comment (dwbuiten) [inline src/bin/muxer/avformat.rs]: This abuses the fact that rav1e's timestamps will always increment by one. Is this OK?

File: src/bin/muxer/avformat.rs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+      if ret < 0 {
+        panic!("write frame failed error = {}", ret);
+      }
+    }
+  }
+
+  fn flush(&mut self) -> io::Result<()> {
+    unsafe {
+      let io = (*self.context).pb;
+      avio_flush(io);
```
Comment (dwbuiten) [inline src/bin/muxer/avformat.rs]: Shouldn't flushing also entail calling `av_write_frame` until it returns 1?

File: src/bin/muxer/mod.rs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+use self::ivf::IvfMuxer;
+
+mod mp4;
+use mp4::Mp4Muxer;
+
+mod y4m;
+pub use self::y4m::write_y4m_frame;
+
+#[cfg(feature = "ffmpeg-sys")]
+mod avformat;
```
Comment (dwbuiten) [inline src/bin/muxer/mod.rs]: What is `avformat` used for in this file?

File: src/bin/rav1e.rs [hunk 13]
```diff
@@ -195,11 +193,23 @@ fn main() {
...
   if video_info.bit_depth == 8 {
     do_encode::<u8, y4m::Decoder<'_, Box<dyn Read>>>(
-      cfg, cli.verbose, progress, err, &mut cli.io.output, &mut source, y4m_enc
+      cfg,
+      cli.verbose,
+      progress,
+      err,
+      &mut *cli.io.output,
+      &mut source,
+      y4m_enc
```
Comment (dwbuiten) [inline src/bin/rav1e.rs]: Unrelated cosmetic changes.

File: src/bin/rav1e.rs [hunk 14]
```diff
@@ -195,11 +193,23 @@ fn main() {
...
   } else {
     do_encode::<u16, y4m::Decoder<'_, Box<dyn Read>>>(
-      cfg, cli.verbose, progress, err, &mut cli.io.output, &mut source, y4m_enc
+      cfg,
+      cli.verbose,
+      progress,
+      err,
+      &mut *cli.io.output,
+      &mut source,
+      y4m_enc
```
Comment (dwbuiten) [inline src/bin/rav1e.rs]: Unrelated cosmetic changes.

File: Cargo.toml [hunk 15]
```diff
@@ -36,6 +36,7 @@ dav1d-sys = { version = "0.2", optional = true }
 aom-sys = { version = "0.1.2", optional = true }
 scan_fmt = { version = "0.2", optional = true }
 ivf = { path = "ivf/", optional = true }
+ffmpeg-sys = { git = "https://github.com/meh/rust-ffmpeg-sys.git", branch = "master", optional = true }
```
Comment (dwbuiten) [inline Cargo.toml]: Is there no real crate?

File: src/bin/muxer/avformat.rs (hunk 7 above)
Comment (xuguangxin) [reply src/bin/muxer/avformat.rs]: for mux, caller need set/close the avio. 

see
https://github.com/FFmpeg/FFmpeg/blob/819ed1df94dff335c2653a83e1d3001aa5797f9e/libavformat/avformat.h#L1374-L1386

File: src/bin/muxer/avformat.rs (hunk 8 above)
Comment (xuguangxin) [reply src/bin/muxer/avformat.rs]: av_write_frame will not take ownership, 
av_init_packet is enough for our case. It's better if we can use uninitialized memory. But I did not find an easy way in rust.

File: src/bin/muxer/avformat.rs (hunk 9 above)
Comment (xuguangxin) [reply src/bin/muxer/avformat.rs]: hmm, this is good catch. Maybe I need to use av_rescale_q. Let me check.

File: src/bin/muxer/avformat.rs (hunk 10 above)
Comment (xuguangxin) [reply src/bin/muxer/avformat.rs]: no, the name is not accurate, you can think duration is time unit. PTS is a number of the unit.
In our special case, one time unit is 1/fps.

File: src/bin/muxer/avformat.rs (hunk 11 above)
Comment (xuguangxin) [reply src/bin/muxer/avformat.rs]: yes, your are right.

File: src/bin/rav1e.rs (hunk 13 above)
Comment (xuguangxin) [reply src/bin/rav1e.rs]: it's formatted by cargo fmt.
I will change it.

File: Cargo.toml (hunk 15 above)
Comment (xuguangxin) [reply Cargo.toml]: No, the crate only works with ffmpeg 3.x, we need  github repo to work with 4.x

**Comment by xuguangxin** [comment]
@dwbuiten , thanks for the review. Seems we do have some issues.
Let me update it in the following days.
thanks

File: Cargo.toml (hunk 15 above)
Comment (lu-zero) [reply Cargo.toml]: This could be problematic. Could you please split the set with the first 2 commit in another PR?

I'd land it immediately and then we can figure out how to get the libavformat muxing sorted out without git crates.

Comment (xuguangxin) [reply Cargo.toml]: sure, wait 2 mins

Comment (xuguangxin) [reply Cargo.toml]: done https://github.com/xiph/rav1e/pull/1304

File: src/bin/muxer/avformat.rs (hunk 8 above)
Comment (dwbuiten) [reply src/bin/muxer/avformat.rs]: 'better'?

File: src/bin/muxer/avformat.rs (hunk 10 above)
Comment (dwbuiten) [reply src/bin/muxer/avformat.rs]: Which is exactly what I said above; it's a special case from the way rav1e handles timestamps. I just want to make sure it's OK to assume this will always be the case.

File: src/bin/muxer/avformat.rs (hunk 8 above)
Comment (xuguangxin) [reply src/bin/muxer/avformat.rs]: I mean the av_init_packet is enough for this case since av_write_frame will not take the packet ownership.
If we use av_packet_alloc, it introduces unnecessary malloc and we need to call av_packet_free at the end of the function. 
Maybe I missed some important things here. Could you explain why you prefer av_packet_alloc?

File: src/bin/muxer/mod.rs (hunk 12 above)
Comment (xuguangxin) [reply src/bin/muxer/mod.rs]: sorry for miss this. the mp4.rs depends on avformat, so  I need this to export avfrormat.rs for mp4.rs

File: src/bin/muxer/avformat.rs (hunk 11 above)
Comment (xuguangxin) [reply src/bin/muxer/avformat.rs]: Just check more code. Actually flush is not so useful for container format.
If we did not call av_write_trailer. the entire stream is not parseable even we flush all things.
Anyway, I will change the code to follow av_write_frame document.

File: src/bin/muxer/avformat.rs (hunk 10 above)
Comment (xuguangxin) [reply src/bin/muxer/avformat.rs]: Not clearly understand the question. Let explain more on the mechanism.
1. mp4 has a mp4_time_scale,  If it's 1000 it means 1s has 1000 mp4 time unit. 
2. rav1e application will give an fps to the muxer. 
3. We use fps as the video_time_scale. 
4. We set duration as mp4_time_scale/video_time_scale. suppose mp4_time_scale is 100, video_time_scale is 25. it means 1 video time unit has 40(1000/25) mp4 time unit.
5. Pts based on video timescale. we use pts*duration to convert it to mp4 timescale.
If we do not want to use fps as video_time_scale. We need to add new params to muxer and change step 3. But I do not think it's necessary, because use fps as the video time scale is perfectly well for video 
thanks

Comment (lu-zero) [reply src/bin/muxer/avformat.rs]: In certain applications could produce a stream with timestamps like this:
```
0 10 25 26 28 29 32 42
```

Right now it does not happen because our input source cannot behave like that.

Once we accept input with arbitrary timestamps we'll have to deal with it.

File: src/bin/muxer/avformat.rs (hunk 8 above)
Comment (dwbuiten) [reply src/bin/muxer/avformat.rs]: Mostly just because it avoids uninitialized memory. I have no strong feelings on this.

File: src/bin/muxer/avformat.rs (hunk 10 above)
Comment (dwbuiten) [reply src/bin/muxer/avformat.rs]: Yes, at some point we will be accepting proper variable framerate input and outputting it as such.

File: Cargo.toml (hunk 15 above)
Comment (xuguangxin) [reply Cargo.toml]: @lu-zero , code rebased.
How to handle this problem?
thanks

**Comment by tdaede** [comment]
Is this ready to be reviewed again?

**Comment by xuguangxin** [comment]
@tdaede , yes, all issues addressed except this one.
I am waiting for feedback from @lu-zero .
thanks 



> No, the crate only works with ffmpeg 3.x, we need github repo to work with 4.x


> This could be problematic. Could you please split the set with the first 2 commit in another PR?
> 
> I'd land it immediately and then we can figure out how to get the libavformat muxing sorted out without git crates.

File: Cargo.toml (hunk 15 above)
Comment (lu-zero) [reply Cargo.toml]: I hadn't time to check which are the crates available. 

Wrapping libavformat isn't this difficult, so worst case we could make the (N+1)th one with just that purpose and put it on crates.io.

Comment (xuguangxin) [reply Cargo.toml]: Hi @lu-zero , thanks for suggestion. Any document/guidance for this?

thanks

Comment (lu-zero) [reply Cargo.toml]: [here](https://rust-lang.github.io/rust-bindgen/tutorial-0.html) the tutorial, [here](https://github.com/rust-av/aom-rs/tree/master/aom-sys) a working example if you want to try yourself. Ping me on irc and I'll try to help.

Comment (xuguangxin) [reply Cargo.toml]: @lu-zero .
Thanks for offering the help. Seems it's not so difficult. I will try it by myself. And ping you if I have a hard issue.

**Comment by xuguangxin** [comment]
@tdaede , I will follow lu-zero's guidance. and start working on it tomorrow. It may need 1~2 weeks to finish the work. 

thanks

**Comment by xuguangxin** [comment]
@lu-zero , after referring to aom-sys. I have made the prototype runnable. Next step is to integrate it to rav1e. I have 2 questions:

1. Seems we have two ways to integrate. 
   1. checkin code to rav1e repo
   2. upload it crates.io.
I prefer i, because ii need cargo.io maintain effort.  We have [many ](https://crates.io/search?q=ffmpeg)ffmpeg packages in crates.io. No one is maintained in the previous half year. I guess if we upload it to cartes.io, we will repeat the same pattern. What is your suggestion?

2. Why name you perfered?  av-sys?


thanks
thanks

**Comment by xuguangxin** [comment]
@lu-zero , done. please help review it.
thanks

File: av-sys/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "av-sys"
```
Comment (lu-zero) [inline av-sys/Cargo.toml]: maybe call it `avformat-sys`

Comment (xuguangxin) [reply av-sys/Cargo.toml]: sure, updated. please help review it again.
thanks

**Comment by xuguangxin** [comment]
@kodabb and @tdaede any feedback?

thanks

File: src/bin/muxer/mp4.rs [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+#[allow(unused_variables)]
+impl Mp4Muxer {
+  pub fn open(path: &str) -> Box<dyn Muxer> {
+    #[cfg(feature = "avformat-sys")] {
+      AvformatMuxer::open(path)
+    }
+    #[cfg(not(feature = "avformat-sys"))] {
+      panic!("need avformat-sys for .mp4, please build with --features=\"avformat-sys\", or you can use .ivf extesion");
```
Comment (tdaede) [inline src/bin/muxer/mp4.rs]: typo: extension

**Comment by xuguangxin** [comment]
@tdaede , thanks, updated.
